### ISSUE TITLE
FusedLayerNormAffineFunctionLegacyBackward is not differentiable twice

### ISSUE THREAD

**Issue by eukaryote31** [NVIDIA/apex#410]
When differentiating twice over a `FusedLayerNorm`, this error is raised.

**Comment by ngimel** [comment]
Unfortunately, FusedLayerNorm in apex is not twice differentiable, to get twice-differentiability you'd need to use LayerNorm from core pytorch.

**Comment by eukaryote31** [comment]
Would it be possible to get this functionality implemented?

ngimel <notifications@github.com> schrieb am Mi., 31. Juli 2019, 17:27:

**Comment by ngimel** [comment]
We will try to upstream our fused layer norm implementation, and that should come with ability to do double backward, but we don't have a timeline for this.

**Comment by eukaryote31** [comment]
I would like to implement this functionality myself because I need it right
away. Any examples of twice-differentiable kernels I could use for
reference?

ngimel <notifications@github.com> schrieb am Mi., 31. Juli 2019, 18:27:

**Comment by ngimel** [comment]
You can have functionality by using torch.nn.LayerNorm, the performance will be worse than FusedLayerNorm. If you want to make FusedLayerNorm in apex double differentiable, you'd need another custom autograd function (similar to what FusedLayerNorm is now https://github.com/NVIDIA/apex/blob/master/apex/normalization/fused_layer_norm.py#L12). The forward of that function should be the backward of FusedLayerNorm, and the backward of this function should be the double backward path. The math for double derivatives is tedious, but it is mostly spelled out in this file in upstream https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/layer_norm_kernel.cpp
